scalafix-core/shared/src/main/scala/scalafix/internal/rule/DisableSyntax.scala

```diff
@@ -35,38 +37,145 @@ final case class DisableSyntax(
+        }
+        val methods = stats.flatMap {
+          case d: Defn.Def => Some(d)
+          case _ => None
+        }
+        if (methods.isEmpty) None else Some(methods)
+      }
+    }
+
+    def hasDefaultArgs(d: Defn.Def): Boolean =
+      d.paramss.exists(_.exists(_.default.isDefined))
+
+    def hasNonImplicitParam(d: Defn.Def): Boolean =
+      d.paramss.exists(_.exists(_.mods.forall(!_.is[Mod.Implicit])))
+
+    ctx.tree.collect {
+      case t @ mod"+" if config.noCovariantTypes =>
+        Seq(
           errorCategory
             .copy(id = "covariant")
             .at(
               "Covariant types could lead to error-prone situations.",
               t.pos
             )
-        case t @ mod"-" if config.noContravariantTypes =>
+        )
+      case t @ mod"-" if config.noContravariantTypes =>
+        Seq(
           errorCategory
             .copy(id = "contravariant")
             .at(
               "Contravariant types could lead to error-prone situations.",
               t.pos
             )
-      }
+        )
+      case t @ WithMethods(methods)
+          if methods.exists(hasDefaultArgs) && config.noDefaultArgs =>
+        methods
+          .filter(hasDefaultArgs)
+          .map(
+            m =>
+              errorCategory
+                .copy(id = "defaultArgs")
+                .at(
+                  "Default args makes it hard to use methods as functions.",
+                  m.pos))
+      case t @ AbstractWithVals(vals) if config.noValInAbstract =>
+        vals.map(
+          v =>
+            errorCategory
+              .copy(id = "valInAbstract")
+              .at(
+                "val definitions in traits/abstract classes may cause initialization bugs",
+                v.pos))
+      case t @ Defn.Object(mods, _, _)
+          if mods.exists(_.is[Mod.Implicit]) && config.noImplicitObject =>
+        Seq(
+          errorCategory
+            .copy(id = "implicitObject")
+            .at("implicit objects may cause implicit resolution errors", t.pos)
+        )
+      case t @ Defn.Def(mods, _, _, paramss, _, _)
+          if mods.exists(_.is[Mod.Implicit]) &&
+            hasNonImplicitParam(t) &&
+            config.noImplicitConversion =>
+        Seq(
+          errorCategory
+            .copy(id = "implicitConversion")
+            .at(
+              "implicit conversions weaken type safety and always can be replaced by explicit conversions",
+              t.pos)
+        )
+    }.flatten
+  }
+
+  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+    checkTree(ctx) ++ checkTokens(ctx) ++ checkRegex(ctx)
+  }
 
-    treeLintMessages ++ tokensLintMessage ++ regexLintMessages.result()
+  override def fix(ctx: RuleCtx): Patch = {
+    ctx.tree.collect {
+      case t @ Defn.Class(mods, _, _, _, _)
+          if config.noNonFinalCaseClass &&
+            mods.exists(_.is[Mod.Case]) &&
+            !mods.exists(_.is[Mod.Final]) =>
+        ctx.addLeft(t, "final ")
```

Comment:
does this handle `@deprecated("blah", "") case class Blah()`? The `final` modifier should appear before `Mod.Case()` but after annotations.

Reply:
See #553 for a similar case where we addLeft on the `val` modifier tree node instead of enclosing `Defn` tree. The scalameta `List[Mod]` API is annoying but it intentionally treats modifiers as tree nodes in order to accommodate cases like this.